src/util/range.h

```diff
@@ -0,0 +1,348 @@
+/*******************************************************************\
+
+Module: Range
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Ranges: pair of begin and end iterators.
+
+#ifndef CPROVER_UTIL_RANGE_H
+#define CPROVER_UTIL_RANGE_H
+
+#include <functional>
+#include <util/make_unique.h>
+
+/// Iterator for which dereference gives the result of some given
+/// function \c f applied on the underlying iterator.
+template <typename iteratort, typename outputt>
+class map_iteratort
+{
+public:
+  using difference_type = void;
+  using value_type = outputt;
+  using pointer = const outputt *;
+  using reference = const outputt &;
+  using iterator_category = std::forward_iterator_tag;
+
+  bool operator==(const map_iteratort &other) const
+  {
+    return underlying == other.underlying;
+  }
+
+  bool operator!=(const map_iteratort &other) const
+  {
+    return !(this->underlying == other.underlying);
+  }
+
+  /// Preincrement operator
+  map_iteratort &operator++()
+  {
+    ++underlying;
+    if(underlying != underlying_end)
+      current = util_make_unique<outputt>((*f)(*underlying));
```

Comment:
⛏ 🤔  I'm not sure this should happen on iteration (maybe on dereference would be better). That could also mean you wouldn't havet to store the result of `current` in the iterator.

Reply:
The dereference operator doesn't have to return const reference, it can also return a value (it's perfectly legal). So in this case, just `return f(underlying)`.

Reply:
What about `operator->`? It has to return a pointer.

Reply:
Maybe you just want to use `mutable`? It seems the documentation doesn't coincide with what is going on here either.